x-pack/filebeat/input/awss3/s3.go

```diff
@@ -250,39 +253,45 @@ func (p *s3Poller) Purge() {
 			// purge calls can overlap, GetListingIDs can return
 			// an outdated snapshot with listing already purged
 			p.states.DeleteListing(listingID)
+			p.log.Debugw("deleting already purged listing from states.", "listingID", listingID)
 			continue
 		}
 
 		lock.(*sync.Mutex).Lock()
 
 		keys := map[string]struct{}{}
-		latestStoredTimeByBucket := make(map[string]time.Time, 0)
+		latestStoredTimeByBucketAndListPrefix := make(map[string]time.Time, 0)
 
 		for _, state := range p.states.GetStatesByListingID(listingID) {
 			// it is not stored, keep
 			if !state.Stored {
+				p.log.Debugw("state not stored, skip purge", "state", state)
 				continue
 			}
 
 			var latestStoredTime time.Time
 			keys[state.ID] = struct{}{}
-			latestStoredTime, ok := latestStoredTimeByBucket[state.Bucket]
+			latestStoredTime, ok := latestStoredTimeByBucketAndListPrefix[state.Bucket+state.ListPrefix]
```

Comment:
this is indeed both a bugfix and a break change
from the elastic side the only integrations that does that is the one on cisco managed s3 buckets, that's indeed the bugged integration that will fix
but in the case of users using custom setup with a single bucket and multiple list prefixes it will be a breaking change as in the measure that s3 objects could be ingested again (and probably fails ingestion because of same document ids)